core/tee/tee_time_generic.c

```diff
@@ -17,50 +21,179 @@ struct tee_ta_time_offs {
 
 static struct tee_ta_time_offs *tee_time_offs;
 static size_t tee_time_num_offs;
+static struct mutex tee_time_mtx = MUTEX_INITIALIZER;
 
-static TEE_Result tee_time_ta_get_offs(const TEE_UUID *uuid,
-				       const TEE_Time **offs, bool *positive)
-{
-	size_t n;
-
-	for (n = 0; n < tee_time_num_offs; n++) {
-		if (memcmp(uuid, &tee_time_offs[n].uuid, sizeof(TEE_UUID))
-				== 0) {
-			*offs = &tee_time_offs[n].offs;
-			*positive = tee_time_offs[n].positive;
-			return TEE_SUCCESS;
-		}
-	}
-	return TEE_ERROR_TIME_NOT_SET;
-}
-
-static TEE_Result tee_time_ta_set_offs(const TEE_UUID *uuid,
-				       const TEE_Time *offs, bool positive)
+static TEE_Result tee_time_ta_set_offs_mem(const TEE_UUID *uuid,
+					   const TEE_Time *offs,
+					   bool positive, size_t *out_idx)
 {
-	size_t n;
-	struct tee_ta_time_offs *o;
+	size_t n = 0;
+	struct tee_ta_time_offs *o = NULL;
 
+	mutex_lock(&tee_time_mtx);
 	for (n = 0; n < tee_time_num_offs; n++) {
-		if (memcmp(uuid, &tee_time_offs[n].uuid, sizeof(TEE_UUID))
-				== 0) {
+		if (memcmp(uuid, &tee_time_offs[n].uuid,
+			   sizeof(TEE_UUID)) == 0) {
 			tee_time_offs[n].offs = *offs;
 			tee_time_offs[n].positive = positive;
+			if (out_idx)
+				*out_idx = n;
+			mutex_unlock(&tee_time_mtx);
 			return TEE_SUCCESS;
 		}
 	}
 
 	n = tee_time_num_offs + 1;
 	o = realloc(tee_time_offs, n * sizeof(struct tee_ta_time_offs));
-	if (!o)
+	if (!o) {
+		mutex_unlock(&tee_time_mtx);
 		return TEE_ERROR_OUT_OF_MEMORY;
+	}
 	tee_time_offs = o;
 	tee_time_offs[tee_time_num_offs].uuid = *uuid;
 	tee_time_offs[tee_time_num_offs].offs = *offs;
 	tee_time_offs[tee_time_num_offs].positive = positive;
 	tee_time_num_offs = n;
+	if (out_idx)
+		*out_idx = n - 1;
+	mutex_unlock(&tee_time_mtx);
 	return TEE_SUCCESS;
 }
 
+static TEE_Result tee_time_ta_storage_write(const TEE_UUID *uuid,
+					    const TEE_Time *offs, bool positive)
+{
+	const struct tee_file_operations *fops =
+		tee_svc_storage_file_ops(TEE_STORAGE_PRIVATE);
+	struct tee_file_handle *fh = NULL;
+	struct tee_pobj *po = NULL;
+	struct tee_ta_time_offs o = { };
+	TEE_Result res = TEE_SUCCESS;
+
+	if (!fops)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	res = tee_pobj_get((TEE_UUID *)uuid, (void *)TA_TIME_OFFS_ID,
```

Comment:
The object will be created in the object namespace for the TA/UUID in question. What if a TA has an object with this name?

I think we can manage with a dummy pobj like in `check_update_version()` instead. This will allow us to create an object outside the normal namespace. However, we must still be careful to avoid conflicts.